core/src/plugins/kubernetes/results-cache-base.ts

```diff
@@ -70,38 +72,46 @@ export class StructuredCacheKey<AdditionalKeyData> {
   }
 }
 
-export class CacheStorageError extends GardenError {
+export abstract class CacheStorageError extends GardenError {
   type = "cache-storage"
+
+  abstract describe(): string
 }
 
-export interface CacheStorage {
+export interface CacheStorage<ResultShape> {
   /**
    * Returns a value associated with the {@code key},
    * or throws a {@link CacheStorageError} if no key was found or any error occurred.
    */
-  get(key: string): Promise<JsonObject>
+  get(key: string, action: Action): Promise<JsonObject>
 
   /**
    * Stores the value associated with the {@code key}.
    *
    * Returns the value back if it was written successfully,
    * or throws a {@link CacheStorageError} otherwise.
    */
-  put(key: string, value: JsonObject): Promise<JsonObject>
+  put(key: string, value: ResultShape, action: Action): Promise<ResultShape>
```

Comment:
It's simpler for this use case to just make the method generic, and not the entire class.
```suggestion
  put<ResultShape>(key: string, value: ResultShape, action: Action): Promise<ResultShape>
```

Reply:
This type must be consistent with the schema injected to the `ResultCache`.
Changing this requires extra refactoring and does not seem to be a blocker. Let's revisit this later.